File: src/shared/sitePatterns.ts
/**
 * Chrome's content-settings pattern grammar, the subset the cookie and site-data lists use
 * (`chrome://settings/content/siteData`, `ContentSettingsPattern`):
 *
 *     [scheme://][*.]host[:port]
 *
 * - `[*.]example.com` covers `example.com` and every subdomain; `example.com` alone covers that
 *   host only. Chrome's "Add site" dialog prepends `[*.]` to a bare host, and so does
 *   {@link sitePatternForHost}: that is what a row added from the site-information sheet gets.
 * - `scheme://` (http, https, ws or wss; `*://` says nothing) and `:port` narrow a pattern to
 *   one scheme or one port; without them a pattern covers every scheme and port. A port cannot
 *   go with a subdomain wildcard in Chrome either.
 * - Hosts are lowercase DNS names (IDN in punycode, as URLs carry them), dotted-decimal IPv4 or
 *   bracketed IPv6 literals; an IP literal takes no `[*.]`. Paths are not part of the grammar.
 *
 * When several patterns of one policy cover a URL the most specific one decides, as Chrome's
 * `ContentSettingsPattern::Compare` orders them: an exact host beats a subdomain wildcard, a
 * longer host beats a shorter one, then a named scheme beats any scheme and a named port beats
 * any port. The Kotlin twin is `privacy/SitePatterns.kt`; pure, no platform code.
 */

import { parseIpv4, parseIpv6 } from './nonUniqueHost'

export interface SitePattern {
  /** The canonical text (`https://[*.]example.com:8443`), what the lists store. */
  text: string
  /** `http`, `https`, `ws` or `wss`; null for any scheme. */
  scheme: string | null
  /** Lowercase host; an IPv6 literal keeps its brackets. */
  host: string
  /** `[*.]`: the host's subdomains are covered too. */
  subdomains: boolean
  /** A fixed port, or null for any. */
  port: number | null
}

const SCHEMES: readonly string[] = ['http', 'https', 'ws', 'wss']
const DEFAULT_PORTS: Record<string, number> = { http: 80, https: 443, ws: 80, wss: 443 }
const HOST_LABEL = /^(?!-)[a-z0-9_-]{1,63}(?<!-)$/
const MAX_HOST_LENGTH = 253
const WILDCARD = '[*.]'

/** Parse `input` as a pattern; null when it is not one. Whitespace and case are forgiven. */
export function parseSitePattern(input: string): SitePattern | null {
  let text = input.trim().toLowerCase()
  if (!text || /[\s/\\?#@]/.test(text)) return null
  let scheme: string | null = null
  const schemeEnd = text.indexOf('://')
  if (schemeEnd !== -1) {
    const given = text.slice(0, schemeEnd)
    text = text.slice(schemeEnd + 3)
    if (given !== '*') {
      if (!SCHEMES.includes(given)) return null
      scheme = given
    }
  }
  let subdomains = false
  if (text.startsWith(WILDCARD)) {
    subdomains = true
    text = text.slice(WILDCARD.length)
  }
  if (!text || text.includes('*')) return null
  let port: number | null = null
  const portMatch = /^(.*?)(?::(\*|\d{1,5}))?$/.exec(text)
  if (!portMatch) return null
  let hostText = portMatch[1]
  if (portMatch[2] !== undefined && portMatch[2] !== '*') {
    port = Number(portMatch[2])
    if (port < 1 || port > 65535) return null
    // A port narrows one host; Chrome refuses the combination with a domain wildcard.
    if (subdomains) return null
  }
  if (hostText.endsWith('.') && hostText.length > 1) hostText = hostText.slice(0, -1)
  const host = canonicalHost(hostText)
  if (!host) return null
  // An IP literal names one machine: nothing is under it.
  if (subdomains && (host.startsWith('[') || parseIpv4(host))) return null
  return { text: patternText(scheme, host, subdomains, port), scheme, host, subdomains, port }
}

/** The canonical text of `input`, or null when it is not a pattern. */
export function normalizeSitePattern(input: string): string | null {
  return parseSitePattern(input)?.text ?? null
}

/** The pattern a bare host gets when added from a page: `[*.]host` (an IP literal stays exact). */
export function sitePatternForHost(host: string): string | null {
  const bare = parseSitePattern(host)
  if (!bare || bare.scheme || bare.port) return null
  if (bare.host.startsWith('[') || parseIpv4(bare.host)) return bare.text
  return `${WILDCARD}${bare.host}`
}

/** The parts of a URL a pattern is matched against. */
export interface SiteAddress {
  scheme: string
  host: string
  port: number | null
}

/** `url` (a URL string, or parts already taken from one) as a {@link SiteAddress}; null for a URL without a host. */
export function siteAddressOf(url: string | SiteAddress): SiteAddress | null {
  if (typeof url !== 'string') return url
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    return null
  }
  const scheme = parsed.protocol.replace(/:$/, '')
  const host = parsed.hostname.toLowerCase().replace(/\.$/, '')
  if (!host) return null
  const port = parsed.port ? Number(parsed.port) : (DEFAULT_PORTS[scheme] ?? null)
  return { scheme, host, port }
}

/** Whether `pattern` covers the URL. */
export function sitePatternMatches(pattern: SitePattern, url: string | SiteAddress): boolean {
  const address = siteAddressOf(url)
  if (!address) return false
  if (pattern.scheme !== null && pattern.scheme !== address.scheme) return false
  if (pattern.port !== null && pattern.port !== address.port) return false
  return hostCovered(pattern, address.host)
}

/** Whether `pattern` covers `host` alone, whatever the scheme or port. */
export function sitePatternCoversHost(pattern: SitePattern, host: string): boolean {
  return hostCovered(pattern, host.toLowerCase().replace(/\.$/, ''))
}

function hostCovered(pattern: SitePattern, host: string): boolean {
  if (host === pattern.host) return true
  if (!pattern.subdomains) return false
  return (
    host.length > pattern.host.length &&
    host.endsWith(pattern.host) &&
    host[host.length - pattern.host.length - 1] === '.'
  )
}

/**
 * Chrome's order of patterns from the most specific to the least: an exact host before a
 * subdomain wildcard, a longer host before a shorter one, a named scheme before any, a named
 * port before any; ties by text so the order is total. Negative when `a` is the more specific.
 */
export function compareSitePatterns(a: SitePattern, b: SitePattern): number {
  if (a.subdomains !== b.subdomains) return a.subdomains ? 1 : -1
  if (a.host.length !== b.host.length) return b.host.length - a.host.length
  if ((a.scheme === null) !== (b.scheme === null)) return a.scheme === null ? 1 : -1
  if ((a.port === null) !== (b.port === null)) return a.port === null ? 1 : -1
  return a.text < b.text ? -1 : a.text > b.text ? 1 : 0
}

/** The most specific of `patterns` (texts; unparsable ones are skipped) covering the URL, or null. */
export function matchSitePatterns(
  patterns: readonly string[],
  url: string | SiteAddress
): SitePattern | null {
  const address = siteAddressOf(url)
  if (!address) return null
  let best: SitePattern | null = null
  for (const text of patterns) {
    const pattern = parseSitePattern(text)
    if (!pattern || !sitePatternMatches(pattern, address)) continue
    if (!best || compareSitePatterns(pattern, best) < 0) best = pattern
  }
  return best
}

/** Whether any of `patterns` covers the URL. */
export function urlInSitePatterns(patterns: readonly string[], url: string | SiteAddress): boolean {
  return matchSitePatterns(patterns, url) !== null
}

function patternText(
  scheme: string | null,
  host: string,
  subdomains: boolean,
  port: number | null
): string {
  return `${scheme ? `${scheme}://` : ''}${subdomains ? WILDCARD : ''}${host}${port ? `:${port}` : ''}`
}

/** A lowercase DNS name, dotted-decimal IPv4 or bracketed IPv6 literal; null for anything else. */
function canonicalHost(text: string): string | null {
  if (!text) return null
  if (text.startsWith('[') && text.endsWith(']'))
    return parseIpv6(text.slice(1, -1)) ? text : null
  if (parseIpv6(text)) return `[${text}]`
  if (/^\d+(\.\d+){3}$/.test(text)) return parseIpv4(text) ? text : null
  if (text.length > MAX_HOST_LENGTH) return null
  const labels = text.split('.')
  if (!labels.every((label) => HOST_LABEL.test(label))) return null
  return text
}

File: android/app/src/main/kotlin/app/zen/chromium/privacy/SitePatterns.kt
package app.zen.chromium.privacy

import app.zen.chromium.blocking.Domains

/**
 * Chrome's content-settings pattern grammar as the cookie and site-data lists use it
 * (`[scheme://][*.]host[:port]`), the twin of `src/shared/sitePatterns.ts`: the same parse, the
 * same match, the same order of specificity, so the header stage's word on a request is the
 * desktop's. `[*.]example.com` covers the host and its subdomains, `example.com` the host alone;
 * a named scheme (`http` / `https`; `*://` says nothing; `ws` / `wss` are matched as the http
 * scheme they ride on) or port narrows a pattern; an IP literal takes no `[*.]`, a port goes with
 * no `[*.]`. Pure, no Android in it.
 */
class SitePattern private constructor(
    /** The canonical text, what the lists store. */
    val text: String,
    /** `http` or `https`; null for any scheme. */
    val scheme: String?,
    /** Lowercase host; an IPv6 literal keeps its brackets. */
    val host: String,
    /** `[*.]`: the host's subdomains are covered too. */
    val subdomains: Boolean,
    /** A fixed port, or null for any. */
    val port: Int?
) : Comparable<SitePattern> {
    /** Whether the pattern covers `address`. */
    fun matches(address: SiteAddress): Boolean {
        if (scheme != null && scheme != address.scheme) return false
        if (port != null && port != address.port) return false
        return coversHost(address.host)
    }

    /** Whether the pattern covers `url`; false for a URL without a host. */
    fun matches(url: String): Boolean = SiteAddress.of(url)?.let { matches(it) } ?: false

    /** Whether the pattern covers `host` alone, whatever the scheme or port. */
    fun coversHost(candidate: String): Boolean {
        val h = candidate.lowercase().removeSuffix(".")
        if (h == host) return true
        if (!subdomains) return false
        return h.length > host.length && h.endsWith(host) && h[h.length - host.length - 1] == '.'
    }

    /**
     * Chrome's order from the most specific to the least (`compareSitePatterns`): an exact host
     * before a subdomain wildcard, a longer host before a shorter one, a named scheme before
     * any, a named port before any; ties by text. Negative when this one is the more specific.
     */
    override fun compareTo(other: SitePattern): Int {
        if (subdomains != other.subdomains) return if (subdomains) 1 else -1
        if (host.length != other.host.length) return other.host.length - host.length
        if ((scheme == null) != (other.scheme == null)) return if (scheme == null) 1 else -1
        if ((port == null) != (other.port == null)) return if (port == null) 1 else -1
        return text.compareTo(other.text)
    }

    override fun equals(other: Any?): Boolean = other is SitePattern && other.text == text
    override fun hashCode(): Int = text.hashCode()
    override fun toString(): String = text

    companion object {
        private const val WILDCARD = "[*.]"
        private val SCHEMES = setOf("http", "https")
        private val HOST_LABEL = Regex("^(?!-)[a-z0-9_-]{1,63}(?<!-)$")
        private val IPV4 = Regex("^\\d+(\\.\\d+){3}$")
        private val PORT = Regex("^(.*?)(?::(\\*|\\d{1,5}))?$")
        private val FORBIDDEN = Regex("[\\s/\\\\?#@]")
        private const val MAX_HOST_LENGTH = 253

        /** Parse `input` as a pattern; null when it is not one. Whitespace and case are forgiven. */
        fun parse(input: String): SitePattern? {
            var text = input.trim().lowercase()
            if (text.isEmpty() || FORBIDDEN.containsMatchIn(text)) return null
            var scheme: String? = null
            val schemeEnd = text.indexOf("://")
            if (schemeEnd != -1) {
                val given = text.substring(0, schemeEnd)
                text = text.substring(schemeEnd + 3)
                if (given != "*") {
                    if (given !in SCHEMES) return null
                    scheme = given
                }
            }
            var subdomains = false
            if (text.startsWith(WILDCARD)) {
                subdomains = true
                text = text.substring(WILDCARD.length)
            }
            if (text.isEmpty() || text.contains('*')) return null
            val portMatch = PORT.matchEntire(text) ?: return null
            var hostText = portMatch.groupValues[1]
            var port: Int? = null
            val portText = portMatch.groups[2]?.value
            if (portText != null && portText != "*") {
                port = portText.toIntOrNull() ?: return null
                if (port < 1 || port > 65535) return null
                // A port narrows one host; Chrome refuses the combination with a domain wildcard.
                if (subdomains) return null
            }
            if (hostText.endsWith(".") && hostText.length > 1) hostText = hostText.dropLast(1)
            val host = canonicalHost(hostText) ?: return null
            // An IP literal names one machine: nothing is under it.
            if (subdomains && (host.startsWith("[") || isIpv4(host))) return null
            return SitePattern(patternText(scheme, host, subdomains, port), scheme, host, subdomains, port)
        }

        /** The canonical text of `input`, or null when it is not a pattern. */
        fun normalize(input: String): String? = parse(input)?.text

        /** The pattern a bare host gets when added from a page: `[*.]host` (an IP literal stays exact). */
        fun forHost(host: String): String? {
            val bare = parse(host) ?: return null
            if (bare.scheme != null || bare.port != null) return null
            if (bare.host.startsWith("[") || isIpv4(bare.host)) return bare.text
            return WILDCARD + bare.host
        }

        /** The most specific of `patterns` (texts; unparsable ones are skipped) covering `address`, or null. */
        fun match(patterns: Iterable<String>, address: SiteAddress): SitePattern? {
            var best: SitePattern? = null
            for (text in patterns) {
                val pattern = parse(text) ?: continue
                if (!pattern.matches(address)) continue
                if (best == null || pattern < best) best = pattern
            }
            return best
        }

        /** The most specific of `patterns` covering `url`, or null (a URL without a host matches nothing). */
        fun match(patterns: Iterable<String>, url: String): SitePattern? =
            SiteAddress.of(url)?.let { match(patterns, it) }

        private fun patternText(scheme: String?, host: String, subdomains: Boolean, port: Int?): String =
            (if (scheme != null) "$scheme://" else "") + (if (subdomains) WILDCARD else "") + host + (if (port != null) ":$port" else "")

        private fun isIpv4(host: String): Boolean = IPV4.matches(host) && NonUniqueHost.parseIpv4(host) != null

        /** A lowercase DNS name, dotted-decimal IPv4 or bracketed IPv6 literal; null for anything else. */
        private fun canonicalHost(text: String): String? {
            if (text.isEmpty()) return null
            if (text.startsWith("[") && text.endsWith("]")) {
                return if (NonUniqueHost.parseIpv6(text.substring(1, text.length - 1)) != null) text else null
            }
            if (NonUniqueHost.parseIpv6(text) != null) return "[$text]"
            if (IPV4.matches(text)) return if (NonUniqueHost.parseIpv4(text) != null) text else null
            if (text.length > MAX_HOST_LENGTH) return null
            val labels = text.split('.')
            if (!labels.all { HOST_LABEL.matches(it) }) return null
            return text
        }
    }
}

/** The parts of a URL a pattern is matched against (`SiteAddress` in `sitePatterns.ts`). */
class SiteAddress(val scheme: String, val host: String, val port: Int?) {
    companion object {
        private val DEFAULT_PORTS = mapOf("http" to 80, "https" to 443)
        private val SCHEME_ALIASES = mapOf("ws" to "http", "wss" to "https")

        /** `url`'s scheme, host and port (the scheme's default without one); null for a URL without a host. */
        fun of(url: String): SiteAddress? {
            val schemeEnd = url.indexOf("://")
            if (schemeEnd <= 0) return null
            val raw = url.substring(0, schemeEnd).lowercase()
            val scheme = SCHEME_ALIASES[raw] ?: raw
            val host = Domains.hostnameOf(url) ?: return null
            var end = url.length
            for (i in schemeEnd + 3 until url.length) {
                val c = url[i]
                if (c == '/' || c == '?' || c == '#') {
                    end = i
                    break
                }
            }
            var authority = url.substring(schemeEnd + 3, end)
            val at = authority.lastIndexOf('@')
            if (at != -1) authority = authority.substring(at + 1)
            val portStart = if (authority.startsWith("[")) authority.indexOf(']') + 1 else 0
            val colon = authority.indexOf(':', portStart)
            val port = if (colon == -1) DEFAULT_PORTS[scheme] else authority.substring(colon + 1).toIntOrNull() ?: DEFAULT_PORTS[scheme]
            return SiteAddress(scheme, host, port)
        }
    }
}
